spec/requests/https_redirects_spec.rb

```diff
@@ -20,32 +24,95 @@
       APP_CONFIG.always_use_ssl = "false"
     end
 
-    it "when request is over HTTP" do
-      get @http_url
-      expect(response.status).to eq 301
-      expect(response.location).to start_with "https://"
+    it "redirects empty path to HTTPS" do
+      expect_redirect(http_url, https_url)
+    end
+
+    it "redirects with path to HTTPS" do
+      expect_redirect("#{http_url}/something", "#{https_url}/something")
+    end
+
+    it "redirects with path and parameter to HTTPS" do
+      expect_redirect("#{http_url}/something?more", "#{https_url}/something?more")
     end
 
     it "does not happen over HTTPS" do
-      get @https_url
+      get https_url
       expect(response.status).to eq 200
     end
 
     it "does not happen when SSL is terminated on proxy" do
-      get @http_url, nil, { "X-Forwarded-Proto" => "https" }
+      get http_url, nil, { "X-Forwarded-Proto" => "https" }
       expect(response.status).to eq 200
     end
   end
 
-  context "when always_use_ssl is false" do
+  shared_context "common no ssl" do
     it "does not happen over HTTP" do
-      get @http_url
+      get http_url
       expect(response.status).to eq 200
     end
 
     it "does not happen over HTTPS" do
-      get @https_url
+      get https_url
+
       expect(response.status).to eq 200
     end
   end
+
+  context "for custom domain" do
+    let!(:domain) { "market.custom.org" }
+
+    before do
+      @community = FactoryGirl.create(:community, :domain => domain, use_domain: true)
+      @community.reload
+    end
+
+    context "when always_use_ssl is true:" do
+
+      include_context "common with ssl"
+
+      it "redirects and keeps additional labels over HTTP" do
+        expect_redirect(www_http_url, www_https_url)
+        expect_redirect(sub_http_url, sub_https_url)
+      end
+    end
+
+    context "when always_use_ssl is false" do
+      include_context "common no ssl"
+    end
+  end
+
+  context "for marketplace subdomain" do
+    let!(:ident)     { "market" }
+    let!(:st_domain) { "example.com" }
+    let!(:domain)    { "#{ident}.#{st_domain}" }
+
+    before do
+      @community = FactoryGirl.create(:community, :ident => ident, use_domain: false)
+      @community.reload
+    end
+
+    before(:each) do
+      @orig_domain = APP_CONFIG.domain
+      APP_CONFIG.domain = st_domain
+    end
+
+    after(:each) do
+      APP_CONFIG.domain = @orig_domain
+    end
+
+    context "when always_use_ssl is true:" do
+      include_context "common with ssl"
+
+      it "redirects and strips additional labels over HTTP" do
+        expect_redirect(www_http_url, https_url)
+        expect_redirect(sub_http_url, https_url)
```

Comment:
The use of `let` variables makes this test a bit hard to read and reason about. By looking at the 2 lines here, I can't fully understand what's happening, without checking the value of these variables. And the definition of these variables are a bit far away. In addition, mixing the lazy-loaded and eager-loaded `let`s and relying on that order requires some extra brain power to understand what values will be used.

I think just hard-coding the URLs here would make this test easier to reason about, but of course, it adds some repetition:

```
expect_redirect("http://something.else.market.example.com", "https://market.example.com")
```